Core/Graphics/Source/Graphics.cpp

```diff
@@ -82,31 +84,62 @@ namespace Babylon
         return m_afterRenderTaskCompletionSource.as_task();
     }
 
+    void Graphics::Impl::EnableRendering()
+    {
+        std::scoped_lock lock{m_bgfxState.Mutex};
+
+        if (!m_bgfxState.Initialized)
+        {
+            // Set the thread affinity (all other rendering operations must happen on this thread).
+            m_renderThreadAffinity = std::this_thread::get_id();
+
+            // Initialize bgfx.
+            auto& init{m_bgfxState.InitState};
+            bgfx::setPlatformData(init.platformData);
+            bgfx::init(init);
+            bgfx::setViewClear(0, BGFX_CLEAR_COLOR | BGFX_CLEAR_DEPTH, 0x443355FF, 1.0f, 0);
+            bgfx::setViewRect(0, 0, 0, static_cast<uint16_t>(init.resolution.width), static_cast<uint16_t>(init.resolution.height));
+            bgfx::touch(0);
+
+            m_bgfxState.Initialized = true;
+            m_bgfxState.Dirty = false;
+
+            // Allow JS awaiters to start doing graphics operations.
+            m_enableRenderTaskCompletionSource.complete();
+        }
+    }
+
+    void Graphics::Impl::DisableRendering()
+    {
+        assert(m_renderThreadAffinity.check());
+
+        std::scoped_lock lock{m_bgfxState.Mutex};
+
+        if (m_bgfxState.Initialized)
+        {
+            bgfx::shutdown();
+            m_bgfxState.Initialized = false;
+            m_enableRenderTaskCompletionSource = {};
+            m_renderThreadAffinity = {};
+        }
+    }
+
     void Graphics::Impl::StartRenderingCurrentFrame()
     {
+        assert(m_renderThreadAffinity.check());
+
         if (m_rendering)
         {
             throw std::runtime_error{"Current frame cannot be started before prior frame has been finished."};
         }
         m_rendering = true;
 
+        // Ensure rendering is enabled
+        EnableRendering();
```

Comment:
nit: As we chatted about offline, if this is exposed publicly such that you _can_ call it, I think I'd prefer that you _have_ to call it, and we assert/fail if a call to render happens while rendering isn't enabled. This applies to calling `DisableRendering` in the destructor, too: off the top of my head, it seems better to make calling that the owner's job with the destructor just asserting/failing if shutting down while rendering is enabled, especially since the lifetime of the Graphics object isn't guaranteed to be managed on the render thread. (This was a problem with the pre-existing call to `bgfx::shutdown`, too, come to think of it.) It's not a deal-breaker for me, but I would definitely prefer that these methods, if they must exist, always be required.

Reply:
Yea, I chatted with @bghgary offline as well and his recommendation was to still do it automatically to not break the current contract. I tend to agree with this, as it simplifies what I believe to be the common case (automatic rendering with no intention of releasing bgfx resources until the app dies). In general, my preference is to optimize for the common case, but allow additional complexity for the uncommon case. That said, I'm willing to defer to you and @bghgary. I guess you two can battle it out? :)
![Fight!](https://media.giphy.com/media/l0HlRziKQeCxkHyWA/giphy.gif)

Reply:
I don't have a strong preference either way.

Reply:
I think it's perfectly fine to go in as-is. We have the discussion and rationale recorded here, so we can always revisit it if we decide we want to change something later.